Internal server error when storing a transaction

**Describe the bug**
An internal server error (500) is triggered sometimes when posting transactions (detected on Optimism chain)

Taking a look at the logs, it seems that is related to the delegates:

```python
2023-10-27 13:00:44,110 [ERROR] [MainProcess] Internal Server Error: /api/v1/safes/0x4ABa01FB8E1f6BFE80c56Deb367f19F35Df0f4aE/multisig-transactions/
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/django/core/handlers/exception.py", line 55, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.10/site-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/local/lib/python3.10/site-packages/django/views/decorators/csrf.py", line 56, in wrapper_view
    return view_func(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/django/views/generic/base.py", line 104, in view
    return self.dispatch(request, *args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/rest_framework/views.py", line 509, in dispatch
    response = self.handle_exception(exc)
  File "/usr/local/lib/python3.10/site-packages/rest_framework/views.py", line 469, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/usr/local/lib/python3.10/site-packages/rest_framework/views.py", line 480, in raise_uncaught_exception
    raise exc
  File "/usr/local/lib/python3.10/site-packages/rest_framework/views.py", line 506, in dispatch
    response = handler(request, *args, **kwargs)
  File "/app/safe_transaction_service/history/views.py", line 692, in post
    serializer.save()
  File "/app/safe_transaction_service/history/serializers.py", line 263, in save
    proposer = SafeContractDelegate.objects.get(
  File "/usr/local/lib/python3.10/site-packages/django/db/models/manager.py", line 87, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/usr/local/lib/python3.10/site-packages/django/db/models/query.py", line 640, in get
    raise self.model.MultipleObjectsReturned(
safe_transaction_service.history.models.SafeContractDelegate.MultipleObjectsReturned: get() returned more than one SafeContractDelegate -- it returned 3!
```

**Additional context**
- Related to https://github.com/5afe/safe-support/issues/315

--- a/safe_transaction_service/__init__.py
+++ b/safe_transaction_service/__init__.py
@@ -1,4 +1,4 @@
-__version__ = "4.27.3"
+__version__ = "4.27.4"
 __version_info__ = tuple(
     int(num) if num.isdigit() else num
     for num in __version__.replace("-", ".", 1).split(".")

--- a/safe_transaction_service/history/models.py
+++ b/safe_transaction_service/history/models.py
@@ -1715,25 +1715,33 @@ class SafeContract(models.Model):
 
 
 class SafeContractDelegateManager(models.Manager):
-    def get_delegates_for_safe(self, address: ChecksumAddress) -> Set[ChecksumAddress]:
-        return set(
-            self.filter(safe_contract_id=address)
-            .values_list("delegate", flat=True)
-            .distinct()
+    def get_for_safe(
+        self, safe_address: ChecksumAddress, owner_addresses: Sequence[ChecksumAddress]
+    ) -> QuerySet["SafeContractDelegate"]:
+        if not owner_addresses:
+            return self.none()
+
+        return self.filter(
+            # If safe_contract is null on SafeContractDelegate, delegates are valid for every Safe
+            Q(safe_contract_id=safe_address)
+            | Q(safe_contract=None)
+        ).filter(delegator__in=owner_addresses)
+
+    def get_for_safe_and_delegate(
+        self,
+        safe_address: ChecksumAddress,
+        owner_addresses: Sequence[ChecksumAddress],
+        delegate: ChecksumAddress,
+    ) -> QuerySet["SafeContractDelegate"]:
+        return self.get_for_safe(safe_address, owner_addresses).filter(
+            delegate=delegate
         )
 
     def get_delegates_for_safe_and_owners(
         self, safe_address: ChecksumAddress, owner_addresses: Sequence[ChecksumAddress]
     ) -> Set[ChecksumAddress]:
-        if not owner_addresses:
-            return set()
         return set(
-            self.filter(
-                # If safe_contract is null on SafeContractDelegate, delegates are valid for every Safe
-                Q(safe_contract_id=safe_address)
-                | Q(safe_contract=None)
-            )
-            .filter(delegator__in=owner_addresses)
+            self.get_for_safe(safe_address, owner_addresses)
             .values_list("delegate", flat=True)
             .distinct()
         )
@@ -1741,7 +1749,8 @@ class SafeContractDelegateManager(models.Manager):
 
 class SafeContractDelegate(models.Model):
     """
-    The owners of the Safe can add users so they can propose/retrieve txs as if they were the owners of the Safe
+    Owners (delegators) can delegate on delegates, so they can propose trusted transactions
+    in their name
     """
 
     objects = SafeContractDelegateManager()
@@ -1751,7 +1760,7 @@ class SafeContractDelegate(models.Model):
         related_name="safe_contract_delegates",
         null=True,
         default=None,
-    )
+    )  # If safe_contract is not defined, delegate is valid for every Safe which delegator is an owner
     delegate = EthereumAddressV2Field()
     delegator = EthereumAddressV2Field()  # Owner who created the delegate
     label = models.CharField(max_length=50)

--- a/safe_transaction_service/history/serializers.py
+++ b/safe_transaction_service/history/serializers.py
@@ -260,9 +260,15 @@ class SafeMultisigTransactionSerializer(SafeMultisigTxSerializerV1):
         if self.validated_data["sender"] in self.validated_data["safe_owners"]:
             proposer = self.validated_data["sender"]
         else:
-            proposer = SafeContractDelegate.objects.get(
-                delegate=self.validated_data["sender"]
-            ).delegator
+            proposer = (
+                SafeContractDelegate.objects.get_for_safe_and_delegate(
+                    self.validated_data["safe"],
+                    self.validated_data["safe_owners"],
+                    self.validated_data["sender"],
+                )
+                .first()
+                .delegator
+            )
 
         multisig_transaction, created = MultisigTransaction.objects.get_or_create(
             safe_tx_hash=safe_tx_hash,

--- a/safe_transaction_service/history/tests/test_models.py
+++ b/safe_transaction_service/history/tests/test_models.py
@@ -1000,11 +1000,11 @@ class TestSafeStatus(TestCase):
         self.assertEqual(safe_status_5.previous(), safe_status_2)
 
 
-class TestSafeContract(TestCase):
-    def test_get_delegates_for_safe(self):
+class TestSafeContractDelegate(TestCase):
+    def test_get_for_safe(self):
         random_safe = Account.create().address
-        self.assertEqual(
-            SafeContractDelegate.objects.get_delegates_for_safe(random_safe), set()
+        self.assertCountEqual(
+            SafeContractDelegate.objects.get_for_safe(random_safe, []), []
         )
 
         safe_contract_delegate = SafeContractDelegateFactory()
@@ -1013,15 +1013,58 @@ class TestSafeContract(TestCase):
         )
         safe_contract_delegate_another_safe = SafeContractDelegateFactory()
         safe_address = safe_contract_delegate.safe_contract.address
+
         self.assertCountEqual(
-            SafeContractDelegate.objects.get_delegates_for_safe(safe_address),
-            [safe_contract_delegate.delegate, safe_contract_delegate_2.delegate],
+            SafeContractDelegate.objects.get_for_safe(
+                safe_address,
+                [safe_contract_delegate.delegator, safe_contract_delegate_2.delegator],
+            ),
+            [safe_contract_delegate, safe_contract_delegate_2],
         )
 
         another_safe_address = safe_contract_delegate_another_safe.safe_contract.address
+        # Use a Safe with an owner not matching
         self.assertCountEqual(
-            SafeContractDelegate.objects.get_delegates_for_safe(another_safe_address),
-            [safe_contract_delegate_another_safe.delegate],
+            SafeContractDelegate.objects.get_for_safe(
+                another_safe_address, [safe_contract_delegate.delegator]
+            ),
+            [],
+        )
+        self.assertCountEqual(
+            SafeContractDelegate.objects.get_for_safe(
+                another_safe_address, [safe_contract_delegate_another_safe.delegator]
+            ),
+            [safe_contract_delegate_another_safe],
+        )
+
+        # Create delegate without Safe
+        safe_contract_delegate_without_safe = SafeContractDelegateFactory(
+            safe_contract=None
+        )
+        self.assertCountEqual(
+            SafeContractDelegate.objects.get_for_safe(
+                safe_address,
+                [
+                    safe_contract_delegate.delegator,
+                    safe_contract_delegate_2.delegator,
+                    safe_contract_delegate_without_safe.delegator,
+                ],
+            ),
+            [
+                safe_contract_delegate,
+                safe_contract_delegate_2,
+                safe_contract_delegate_without_safe,
+            ],
+        )
+        self.assertCountEqual(
+            SafeContractDelegate.objects.get_for_safe(
+                another_safe_address,
+                [
+                    safe_contract_delegate_another_safe.delegator,
+                    safe_contract_delegate_without_safe.delegator,
+                ],
+            ),
+            [safe_contract_delegate_another_safe, safe_contract_delegate_without_safe],
         )
 
     def test_get_delegates_for_safe_and_owners(self):

--- a/safe_transaction_service/history/tests/test_views.py
+++ b/safe_transaction_service/history/tests/test_views.py
@@ -1677,11 +1677,16 @@ class TestViews(SafeTestCaseMixin, APITestCase):
             response.data["non_field_errors"][0],
         )
 
-        # Add delegate
-        SafeContractDelegateFactory(
+        # Add delegates (to check there's no issue with delegating twice to the same account)
+        safe_contract_delegate = SafeContractDelegateFactory(
             safe_contract__address=safe_address,
             delegate=safe_delegate.address,
             delegator=safe_owners[0].address,
+        )
+        SafeContractDelegateFactory(
+            safe_contract=safe_contract_delegate.safe_contract,
+            delegate=safe_delegate.address,
+            delegator=safe_owners[1].address,
         )
         response = self.client.post(
             reverse("v1:history:multisig-transactions", args=(safe_address,)),
@@ -1967,9 +1972,13 @@ class TestViews(SafeTestCaseMixin, APITestCase):
         self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
 
         self.assertEqual(SafeContractDelegate.objects.count(), 2)
+        queryset = SafeContractDelegate.objects.get_for_safe(
+            safe_address, [delegator.address]
+        )
+        self.assertEqual(len(queryset), 2)
         self.assertCountEqual(
-            SafeContractDelegate.objects.get_delegates_for_safe(safe_address),
-            [delegate.address],
+            set(safe_contract_delegate.delegate for safe_contract_delegate in queryset),
+            {delegate.address},
         )
 
     def test_delegates_get(self):

